x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/AnalyticsCollectionService.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.entsearch.analytics;
+
+import org.elasticsearch.ResourceAlreadyExistsException;
+import org.elasticsearch.ResourceNotFoundException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.datastreams.CreateDataStreamAction;
+import org.elasticsearch.action.datastreams.DeleteDataStreamAction;
+import org.elasticsearch.action.support.IndicesOptions;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.client.internal.OriginSettingClient;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.service.ClusterService;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ENT_SEARCH_ORIGIN;
+
+/**
+ * A service that allows the manipulation of persistent {@link AnalyticsCollection} model.
+ *
+ * Until we have more specific need the {@link AnalyticsCollection} is just another representation
+ * of a {@link org.elasticsearch.cluster.metadata.IndexAbstraction.DataStream}.
+ *
+ * As a consequence, this service is mostly a facade for the data stream API.
+ */
+public class AnalyticsCollectionService implements ClusterStateListener {
+
+    private final Client clientWithOrigin;
+
+    private final IndexNameExpressionResolver indexNameExpressionResolver;
+
+    private Map<String, AnalyticsCollection> analyticsCollections = Collections.emptyMap();
+
+    public AnalyticsCollectionService(
+        Client client,
+        ClusterService clusterService,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        this.clientWithOrigin = new OriginSettingClient(client, ENT_SEARCH_ORIGIN);
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+        clusterService.addListener(this);
```

Comment:
Should we initialise the analytics collection map here? I am not sure when the cluster changed listener kicks in.

Reply:
I tried it but it seems that the cluster state is not ready when the service is instantiated in the plugin createComponents method. Because of it the following exception is raised:

```
java.lang.AssertionError: initial cluster state not set yet
	at org.elasticsearch.cluster.service.ClusterApplierService.state(ClusterApplierService.java:180)
	at org.elasticsearch.cluster.service.ClusterService.state(ClusterService.java:141)
	at org.elasticsearch.xpack.entsearch.analytics.AnalyticsCollectionService.<init>(AnalyticsCollectionService.java:56)
```

I have been thinking about this more and for me, it makes sense. If you look at the `IndexTemplateRegistry` initialize method, it is just registering itself as a cluster listener.

Reply:
TL:DR: I don't think wwe have something to do.